docs/Developers/Design/CRD_design.md

```diff
@@ -0,0 +1,438 @@
+
+The `spec.management` sub-tree determines how Gluster nodes within this zone
+are managed. `management.mode` chooses one of the three possible management
+modes:
+
+- `automatic`: When in automatic mode, the operator adds and removes Gluster
+  pods as needed for capacity and performance reasons, subject to the
+  restrictions listed in `management.automatic.*`. These restrictions take the
+  form of constraining the total number of pods between a minimum and maximum
+  (`minNodes` and `maxNodes`), limiting the maximum amount of raw capacity that
+  can be allocated (`maxStorage`), and providing capacity thresholds for the
+  operator to use when dynamically adjusting the number of pods
+  (`freeStorageMin`, `freeStorageMax`).
+- `external`: The external mode is used when the zone represents nodes that are
+  not part of the kubernetes cluster (i.e., Gluster servers running in
+  "independent" mode). In this mode, the `management.external.*` subtree is
+  used. Here, `external.nodes` provides a list of the IP/DNS addresses of the
+  individual cluster nodes in this zone. It is also possible to supply ssh
+  credentials via `external.credentials` for use by Heketi's ssh executor to
+  contact glusterd.
+- `manual`: The manual mode is used as an alternative to the automatic mode,
+  when it is desirable to directly specify the number of storage nodes within a
+  zone. The number is directly set via `manual.nodes`.
+
+The `spec.nodeAffinity`, applicable to both automatic and manual mode, is used
+to constrain which nodes the Gluster pods may be placed on. The structure of
+this field directly mirrors the Kubernetes pod's `spec.affinity.nodeAffinity`.
+
+The final section of the Zone spec is the `storage` sub-tree. This set of
+parameters determines how the storage for the pods in this zone are obtained.
+There are two main modes:
+
+- `device` mode: In this mode, the physical devices on the host are listed
+  here, and the pod is given direct access to the host's physical disks. The
+  `storage.device` list provides the device names that should be used.
+- `volume` mode: In this mode, block-mode PVs are assigned to the Gluster pods.
+  The `storage.volume.size` and `storage.volume.storageClassName` provide the
+  information necessary to obtain PersistentVolumes for use in this zone.
+
+## Node CR
+
+The Node CR's usage is slightly different than the Cluster and Zone objects.
+The Cluster and Zone objects are created by the administrator when configuring
+the cluster, but the Node objects are created and deleted dynamically by the
+operator.
+
+```yaml
+apiVersion: "operator.gluster.org/v1alpha1"
+kind: Node
+metadata:
+  # Name for this node
+  name: az1-001
+  # Nodes belong to a zone
+  zone: az1
+  # Nodes belong to a cluster also
+  cluster: my-cluster
+  # CRD is namespaced
+  namespace: gluster
+spec:
+  # Admin (or operator) sets desired state for the node.
+  targetState: active # (active | quiesce)
```

Comment:
Why 2 states here ? cant operator and admin co-ordinate on one state?

Reply:
The distinction between desired state and actual state is important. Just because the admin wants to mark the node as active doesn't mean it's actually available immediately.